File: src/main/java/org/modelingvalue/nelumbo/patterns/NodeTypePattern.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.patterns;

import java.io.Serial;
import java.util.function.Function;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Map;
import org.modelingvalue.collections.mutable.MutableList;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.ConstructionReason;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.NelumboConstructor;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.NodeInfo;
import org.modelingvalue.nelumbo.lang.Functor;
import org.modelingvalue.nelumbo.lang.Type;
import org.modelingvalue.nelumbo.lang.Variable;
import org.modelingvalue.nelumbo.syntax.ParseContext;
import org.modelingvalue.nelumbo.syntax.ParseException;
import org.modelingvalue.nelumbo.syntax.ParseState;
import org.modelingvalue.nelumbo.syntax.Token;
import org.modelingvalue.nelumbo.syntax.TokenType;

public class NodeTypePattern extends Pattern {
    @Serial
    private static final long serialVersionUID = 6828401544789430678L;

    public NodeTypePattern(Type type, List<AstElement> elements, Type nodeType, Integer precedence, Boolean visible) {
        super(NodeInfo.of(type, elements), visible, nodeType, precedence);
    }

    @NelumboConstructor
    public NodeTypePattern(NodeInfo nodeInfo, Object... args) {
        super(nodeInfo, args);
    }

    @Override
    protected NodeTypePattern set(NodeInfo nodeInfo, Object[] args) {
        return new NodeTypePattern(nodeInfo, args);
    }

    public Type nodeType() {
        return (Type) get(1);
    }

    @Override
    public Variable variable() {
        return nodeType().variable();
    }

    @Override
    public NodeTypePattern set(int i, Object... a) {
        return (NodeTypePattern) super.set(i, a);
    }

    @Override
    protected NodeTypePattern setBinding(Node declaration, Map<Variable, Object> vars, boolean setFunctorOrType) {
        Variable var = variable();
        if (var != null && vars.get(var) instanceof Type type) {
            return set(1, nodeType().rewrite(type));
        }
        return (NodeTypePattern) super.setBinding(declaration, vars, setFunctorOrType);
    }

    public Integer precedence() {
        return (Integer) get(2);
    }

    public Boolean visible() {
        return (Boolean) get(0);
    }

    @Override
    public ParseState state(ParseState next) {
        Boolean visible = visible();
        return new ParseState(nodeType(), precedence(), visible != null ? next.setVisibility(visible) : next);
    }

    @Override
    public String toString(TokenType[] previous) {
        return "<" + nodeType() + ">";
    }

    @Override
    public Pattern setPresedence(int precedence) {
        if (precedence() != null) {
            return this;
        }
        return set(2, precedence);
    }

    @Override
    public Pattern setTypes(Function<Type, Type> typeFunction) {
        return set(1, typeFunction.apply(nodeType()));
    }

    @Override
    public List<Type> argTypes(List<Type> types) {
        return types.add(nodeType());
    }

    @Override
    public Pattern setIsConnected() {
        return this;
    }

    @Override
    public int string(List<Object> args, int ai, StringBuffer sb, TokenType[] previous, boolean alt) {
        if (ai < 0 || args.size() <= ai) {
            return -1;
        }
        if (args.get(ai) instanceof Node node && (//
        nodeType().isAssignableFrom(node instanceof Type type ? type : node.type()) || //
                (node instanceof Variable && nodeType().isAssignableFrom(Type.VARIABLE)))) {
            boolean parenthetical = false;
            Functor functor = node.functor();
            ParseState post = functor != null ? functor.postStart() : null;
            if (post != null) {
                Integer inner = precedence();
                parenthetical = inner != null && inner > post.leftPrecedence();
            }
            if (parenthetical) {
                sb.append('(');
            }
            sb.append(node.toString(previous));
            if (parenthetical) {
                sb.append(')');
            }
            return ai + 1;
        }
        return -1;
    }

    @Override
    public int args(List<AstElement> elements, int i, MutableList<Object> args, boolean alt, Functor functor,
            Map<Variable, Type> typeArgs) {
        if (i < elements.size()) {
            AstElement e = elements.get(i);
            if (e instanceof Node n) {
                Type type = nodeType();
                if (type.isAssignableFrom(n.type())) {
                    args.add(n);
                    return i + 1;
                } else if (Type.VARIABLE.equals(type) && n instanceof Variable) {
                    args.add(n);
                    return i + 1;
                } else {
                    Variable var = type.variable();
                    type = var != null ? typeArgs.get(var) : null;
                    if (type != null && type.isAssignableFrom(n.type())) {
                        args.add(n);
                        return i + 1;
                    }
                }
            }
        }
        return -1;
    }

    @Override
    public Pattern declaration(Token token) {
        return null;
    }

    @Override
    public Node init(KnowledgeBase knowledgeBase, ParseContext ctx, ConstructionReason reason) throws ParseException {
        if (reason == ConstructionReason.parsing) {
            List<AstElement> elements = astElements();
            Boolean visible = null;
            String v = getVal(0);
            if (v != null) {
                visible = v.equals("visible");
            }
            Type type = (Type) get(1);
            Integer precedence = null;
            String p = getVal(2);
            if (p != null) {
                precedence = Integer.parseInt(p);
            }
            TokenType tt = type.tokenType();
            return tt != null ? t(elements, tt) : n(elements, type, precedence, visible);
        }
        return this;
    }
}

File: web-ts/src/patterns/NodeTypePattern.ts
/**
 * NodeTypePattern - matches a node of a specific type.
 * @JAVA_REF org.modelingvalue.nelumbo.patterns.NodeTypePattern
 */

import { List, Map } from 'immutable';
import { TokenType } from '../syntax/TokenType';
import type { Token } from '../syntax/Token';
import type { AstElement } from '../AstElement';
import { Type } from '../Type';
import { Variable } from '../Variable';
import { Node } from '../Node';
import { Pattern } from './Pattern';
import { ParseState } from '../syntax/ParseState';
import type { Functor } from './Functor';

export class NodeTypePattern extends Pattern {
  constructor(type: Type, elements: List<AstElement>, ...args: unknown[]) {
    super(type, elements, ...args);
  }

  protected static fromData(data: unknown[], declaration?: Node): NodeTypePattern {
    const pattern = Object.create(NodeTypePattern.prototype) as NodeTypePattern;
    (pattern as unknown as { _data: unknown[] })._data = data;
    (pattern as unknown as { _declaration: Node })._declaration = declaration ?? pattern;
    return pattern;
  }

  protected struct(data: unknown[], declaration?: Node): NodeTypePattern {
    return NodeTypePattern.fromData(data, declaration ?? this.declaration());
  }

  nodeType(): Type {
    return this.get(0) as Type;
  }

  variable(): Variable | null {
    return this.nodeType().variable();
  }

  precedence(): number | null {
    return this.get(1) as number | null;
  }

  setBinding(vars: Map<Variable, unknown>, declaration?: Node): NodeTypePattern {
    const v = this.variable();
    if (v !== null) {
      const val = vars.get(v);
      if (val instanceof Type) {
        return this.set(0, this.nodeType().rewrite(val)) as NodeTypePattern;
      }
    }
    return super.setBinding(vars, declaration) as NodeTypePattern;
  }

  // @JAVA_REF NodeTypePattern.state(ParseState next)
  parseState(next: ParseState): ParseState {
    return new ParseState(this.nodeType(), next, this.precedence());
  }

  toString(_previous?: TokenType[]): string {
    return '<' + this.nodeType().name() + '>';
  }

  setPrecedence(precedence: number): Pattern {
    if (this.precedence() !== null) {
      return this;
    }
    return this.set(1, precedence) as NodeTypePattern;
  }

  setTypes(typeFunction: (type: Type) => Type): Pattern {
    return this.set(0, typeFunction(this.nodeType())) as NodeTypePattern;
  }

  argTypes(types: List<Type>): List<Type> {
    return types.push(this.nodeType());
  }

  string(args: List<unknown>, ai: number, sb: string[], previous: TokenType[], _alt: boolean): number {
    const arg = args.get(ai);
    if (arg instanceof Node) {
      const node = arg;
      const nodeT = node.type();

      // Check if type is assignable or if it's a Variable matching VARIABLE type
      if (this.nodeType().isAssignableFrom(nodeT) ||
          (node instanceof Variable && this.nodeType().isAssignableFrom(Type.VARIABLE))) {

        let parenthetical = false;
        const nodeFunctor = node.functor();
        const post = nodeFunctor?.postStart?.() ?? null;

        if (post !== null) {
          const inner = this.precedence();
          if (inner !== null && inner > (post.leftPrecedence() ?? Number.MAX_SAFE_INTEGER)) {
            parenthetical = true;
          }
        }

        if (parenthetical) {
          sb.push('(');
        }
        sb.push(node.toString(previous));
        if (parenthetical) {
          sb.push(')');
        }
        return ai + 1;
      }
    }
    return -1;
  }

  extractArgs(
    elements: List<AstElement>,
    i: number,
    args: unknown[],
    _alt: boolean,
    _functor: Functor,
    typeArgs: Map<Variable, Type>
  ): number {
    if (i < elements.size) {
      const e = elements.get(i);
      // In Java, Variable extends Node so instanceof Node covers both.
      // In TS, Variable is separate, so we handle both explicitly.
      if (e instanceof Node || e instanceof Variable) {
        const n = e;
        const nType = e instanceof Node ? (e as Node).type() : (e as Variable).type();
        const nodeType = this.nodeType();

        if (nodeType.isAssignableFrom(nType)) {
          args.push(n);
          return i + 1;
        } else if (Type.VARIABLE.equals(nodeType) && e instanceof Variable) {
          args.push(n);
          return i + 1;
        } else {
          const v = nodeType.variable();
          if (v !== null) {
            const resolvedType = typeArgs.get(v);
            if (resolvedType !== undefined && resolvedType.isAssignableFrom(nType)) {
              args.push(n);
              return i + 1;
            }
          }
        }
      }
    }
    return -1;
  }

  tokenDeclaration(_token: Token): Pattern | null {
    return null;
  }
}
